### PR TITLE
Document that Symbol.Sident flexible array member-like semantics

### PR DESCRIPTION
**Pull request by Geod24**
As I was working on unrelated change, I was held back by the inability to change things because of the reliance on C++-style coding practice. Additionally, the extremely unsafe usage of `Symbol.Sident` was a surprise, so I added a comment to document it.

### CODE REVIEW

**Comment by WalterBright** [comment]
The backend is designed using the "separate compilation" model. This is deliberate. All this change does is slow down the edit/compile/debug cycle.

> the inability to change things

Need a more specific rationale.

**Comment by Geod24** [comment]
> Need a more specific rationale.

This was the commit that lead me to this: https://github.com/Geod24/dmd/commit/9b4725ea00fab86b5681deceb12c067952d2cf68

I can't make those `extern(D)` in the header because, as mentioned, module mangling. I can use import, but it has to be selective import otherwise the module conflicts (and if the module importing uses both overloads, game over).

**Comment by WalterBright** [comment]
> I can't make those extern(D) in the header because, as mentioned, module mangling.

Just make it an inline function in the header file. There are a couple others in cc.d.

**Comment by Geod24** [comment]
> Just make it an inline function in the header file. There are a couple others in cc.d.

Sorry, wrong wording. Not header, but module. I can't declare the function as `extern(D)` in `globals` because then it would try to link in `dmd.backend.globals.symbol_calloc` instead of `dmd.backend.symbol.symbol_calloc`.

**Comment by WalterBright** [comment]
You can make a symbol.di file.

File: src/dmd/backend/cc.d [hunk 1]
```diff
@@ -1436,7 +1436,14 @@ struct Symbol
...
 
-    char[1] Sident;             // identifier string (dynamic array)
+    /**
+     * Identifier for this symbol
+     *
+     * Note that this is used as a flexible array member.
+     * When allocating a Symbol, the allocation is for
+     * `sizeof(Symbol - 1 + strlen(identifier) + "\0".length)`.
+     */
+    char[1] Sident;
```
Comment (thewilsonator) [inline src/dmd/backend/cc.d:1446]: Isn't `char[0]` suppose to be use for that?

**Comment by Geod24** [comment]
That doesn't make any sense. One, because the backend is compiled all at once, hence the cost of the extra import should be negligible, if it even exists, because `importAll` will already find it in the root modules (as it did before).

Second, this is pushing complexity on the user. If we have to start using `di` files because template-less modules are too slow to import, we have completely failed. But honestly... They're not.

In any case, I don't want to die on that hill, so I simply removed the last commit, leaving in documentation and style change only, which should be non-controversial.

File: src/dmd/backend/cc.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/backend/cc.d:1446]: I didn't write that code, I am merely documented its usage 🤷‍♂

**Comment by Geod24** [comment]
Ping, this is trivial

**Review by thewilsonator — APPROVED** [review]

**Comment by MoonlightSentinel** [comment]
Seems like DlangBot is on vacation again

**Pull request merged**